dispatcher_startup.go

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package yarpc
+
+import (
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/errorsync"
+	"go.uber.org/zap"
+)
+
+type starter struct {
+	startedMu sync.Mutex
+	started   []transport.Lifecycle
+
+	dispatcher *Dispatcher
+	log        *zap.Logger
+}
+
+func (s *starter) start(lc transport.Lifecycle) func() error {
+	return func() error {
+		if lc == nil {
+			return nil
+		}
+
+		if err := lc.Start(); err != nil {
+			return err
+		}
+
+		s.startedMu.Lock()
+		s.started = append(s.started, lc)
+		s.startedMu.Unlock()
+
+		return nil
+	}
+}
+
+func (s *starter) abort(errs []error) error {
+	// Failed to start so stop everything that was started.
+	wait := errorsync.ErrorWaiter{}
+	s.startedMu.Lock()
+	for _, lc := range s.started {
+		wait.Submit(lc.Stop)
+	}
+	s.startedMu.Unlock()
+	if newErrors := wait.Wait(); len(newErrors) > 0 {
+		errs = append(errs, newErrors...)
+	}
+
+	return multierr.Combine(errs...)
+}
+
+func (s *starter) setRouters() {
+	// don't need a sync.Once, since we call this in a lifecycle.Once in the
+	// dispatcher.
+	s.log.Debug("setting router for inbounds")
+	for _, ib := range s.dispatcher.inbounds {
+		ib.SetRouter(s.dispatcher.table)
+	}
+	s.log.Debug("set router for inbounds")
+}
+
+func (s *starter) startTransports() error {
+	s.log.Info("starting transports")
+	wait := errorsync.ErrorWaiter{}
+	for _, t := range s.dispatcher.transports {
+		wait.Submit(s.start(t))
+	}
+	if errs := wait.Wait(); len(errs) != 0 {
+		return s.abort(errs)
+	}
+	s.log.Debug("started transports")
+	return nil
+}
+
+func (s *starter) startOutbounds() error {
+	s.log.Info("starting outbounds")
+	wait := errorsync.ErrorWaiter{}
+	for _, o := range s.dispatcher.outbounds {
+		wait.Submit(s.start(o.Unary))
+		wait.Submit(s.start(o.Oneway))
+		wait.Submit(s.start(o.Stream))
```

Comment:
If the outbound doesn't support one or more of these, `s.start(nil)` will
return `nil`. `wait.Submit` will try to call the `nil` function as-is.

https://github.com/yarpc/yarpc-go/blob/52507626b429effff0b66de8ed238d87b279b13a/internal/errorsync/err.go#L40

Reply:
Really? I intended `s.start(nil)` to return a non-nil `func() error` that always returns a `nil` error. I think I've done that (above, in this file) - the logic is nearly identical to the existing code.

Reply:
Update, the `s.start` implementation is now below these comments.

Reply:
derp. you're right.